### ISSUE TITLE
spanner: data is being persisted when error is returned

### ISSUE THREAD

**Issue by Jarema** [googleapis/google-cloud-go#1781]
**Client**

Spanner

**Environment**

GKE VM's and Kubernetes (with Debian Slim images)

**Code**

e.g.
```go
_, err := client.Apply(ctx, []*spanner.Mutation{
    spanner.Insert("Users",
        []string{"name", "email"},
        []interface{}{"alice", "a@example.com"})})
if err != nil {
    // TODO: Handle error.
}
row, err := client.Single().ReadRow(ctx, "Users",
    spanner.Key{"alice"}, []string{"email"})
if err != nil {
    error is Context canceled
}
```

**Expected behavior**

When error other than  `Transaction outcome unknown`  is returned, no data is persisted in Database

**Actual behavior**

Spanner sometimes persist data if error is `Context cancel` which suggests, that client when it encounter context cancel does not check if it already send commit to database or not.
In cases when it did send commit I would expect `Transaction outcome unknown`.

**Additional context**

client:
cloud.google.com/go/spanner v1.1.0

PS This is a huge issue impacting our services response times, as we need to manually check if data was persisted for every error from Spanner client. That takes a lot time.

**Comment by olavloite** [comment]
@Jarema Just to be sure that I correctly understand what you are saying:
1. You would like the `Apply` method (and thereby also the `ReadWriteTransaction` method) to return a specific `Transaction outcome unknown` error if a `Context canceled` or `Context timed out` error occurs after a `Commit` has been sent to the backend.
1. All other errors (both `Context` errors and other errors) should remain as they currently are.

This way you would know when to specifically check whether transaction succeeded or not. Is this correct?

The reason I want to make sure is that you wrote:
> PS This is a huge issue impacting our services response times, as we need to manually check if data was persisted for every error from Spanner client. That takes a lot time.

My assumption is that the above quote is missing '... for every _`context`_ error from Spanner client. ...'. In other words: You currently need to check whether the transaction succeeded for all context errors. Other errors, such as for example an `INVALID_ARGUMENT` error does not need to be checked, as it is clear that the transaction did not succeed.

**Comment by Jarema** [comment]
@olavloite Exactly. Apply was just en exaply. It applies to all `ReadWriteTransaction`

Because we're working in very sensitive environment (banking), when we saw that data is persisted for errors other than `Transaction outcome unknown`, we decided to check the Database state after every error, which is ugly code and practice.

If we could be sure that data **might** be persisted **only** if `Transaction outcome unknown` error is returned, it would help us a lot.

Also, that should be documented. Right now I could not find docs about errors and possible outcomes when those happen.
